fs/fs.go

```diff
@@ -81,13 +82,66 @@ var (
 	fusermountBin               = "fusermount"
 )
 
+type layerQueue struct {
+	mu       sync.Mutex
+	queue    chan func() string
+	cache    *hashset.Set
+	maxSize  int
+	currSize int
+}
+
+func (lq *layerQueue) Start() error {
+	if lq.queue != nil {
+		return fmt.Errorf("layer queue already instantiated")
+	}
+
+	lq.queue = make(chan func() string)
+	lq.cache = hashset.New()
+
+	go func() {
+		for {
+			// Wait for queue to be freed
+			for lq.currSize >= lq.maxSize && lq.maxSize != 0 {
+				time.Sleep(100 * time.Millisecond)
+			}
+			lq.mu.Lock()
+			lq.currSize++
+			lq.mu.Unlock()
+
+			preresolveFn := <-lq.queue
+			go func() {
+				digest := preresolveFn()
+
+				lq.mu.Lock()
+				lq.currSize--
+				lq.cache.Remove(digest)
+				lq.mu.Unlock()
+			}()
+		}
+	}()
+
+	return nil
+}
+
+func (lq *layerQueue) Queue(digest string, fn func() string) {
+	if !lq.cache.Contains(digest) {
+		lq.queue <- fn
```

Comment:
nit: we should probably do a non-blocking write here so that we don't get stuck waiting to setup pre-resolving if the snapshotter falls behind.